### PR TITLE
internal: Use `triomphe::Arc`

### LINKED ISSUE #14717: Consider replacing `std::sync::Arc` usages with an `Arc` implementation that has no weak support
We use `Arc`s a lot, but never make use of the weak count so it might be beneficial to use a slimmer Arc (for our attribute handling it might even be more beneficial to employ a `ThinArc`).

### PR DESCRIPTION
**Pull request by lnicola**
Closes #14717

### CODE REVIEW

File: crates/hir-expand/src/attrs.rs [hunk 1]
```diff
@@ -78,8 +81,9 @@ impl RawAttrs {
                                         << AttrId::AST_INDEX_BITS;
                                 it
                             }))
-                            .collect(),
-                    ),
+                            // FIXME: use `Arc::from_iter` when it becomes available
```
Comment (lnicola) [inline crates/hir-expand/src/attrs.rs:84]: Still need to fix these up.

**Comment by Veykril** [comment]
Nice to see that it mostly just works as a drop-in replacement

**Comment by lnicola** [comment]
> Nice to see that it mostly just works as a drop-in replacement

:grimacing: 

Think we can merge this now? It might tend to bitrot. I can fix the FIXMEs when the next `triomphe` release gets out.

We could also use `ThinArc` in some places (there's at least one FIXME for it), but I'd rather do that in another PR.

**Comment by Veykril** [comment]
We are forced to use std's Arc by chalk in some places I believe, other than that once we can make use of clippy properly on this repo we can forbid the std one to be used. I think it should be fine then.

**Review by Veykril — APPROVED** [review]

**Comment by lnicola** [comment]
And `ThinArc` might bring some extra wins.

**Comment by HKalbasi** [comment]
> other than that once we can make use of clippy properly on this repo we can forbid the std one to be used

hmm, then it would be good enough I think.

> Also of note regarding memory usage. 4Mb might be more when IDE features have been used, since analysis stats only really tests type inference + body lowering memory usage and not much more.

Maybe we can measure it by using lsif instead of analysis stats?

**Comment by lnicola** [comment]
```
# baseline
76.42user 0.78system 1:17.47elapsed 99%CPU (0avgtext+0avgdata 1692104maxresident)k

# pr
74.30user 0.68system 1:15.27elapsed 99%CPU (0avgtext+0avgdata 1688284maxresident)k
```

**Comment by Veykril** [comment]
I don't think lsif will test much else, that also just typechecks the project to resolve definitions.

**Comment by HKalbasi** [comment]
It also executes hover on everything. Do you mean an specific ide feature, or the general usage within ide (changing files, ...)?

**Comment by Veykril** [comment]
general usage over time for ide features ye

**Pull request merged**